diff -up shadow-4.9/lib/shadowlog.c.debug3 shadow-4.9/lib/shadowlog.c
--- shadow-4.9/lib/shadowlog.c.debug3	2022-01-10 11:16:31.636261531 +0100
+++ shadow-4.9/lib/shadowlog.c	2022-01-10 11:16:31.637261538 +0100
@@ -2,8 +2,8 @@
 
 #include "lib/shadowlog_internal.h"
 
-const char *shadow_progname;
-FILE *shadow_logfd;
+const char *shadow_progname = "libshadow";
+FILE *shadow_logfd = NULL;
 
 void log_set_progname(const char *progname)
 {
diff -up shadow-4.9/libsubid/api.c.debug3 shadow-4.9/libsubid/api.c
--- shadow-4.9/libsubid/api.c.debug3	2022-01-10 11:16:31.637261538 +0100
+++ shadow-4.9/libsubid/api.c	2022-01-10 11:17:15.431574120 +0100
@@ -40,17 +40,16 @@
 #include "subid.h"
 #include "shadowlog.h"
 
-const char *Prog = "(libsubid)";
-
 bool libsubid_init(const char *progname, FILE * logfd)
 {
 	FILE *shadow_logfd;
 	if (progname) {
 		progname = strdup(progname);
-		if (progname)
-			Prog = progname;
-		else
+		if (!progname)
 			return false;
+		log_set_progname(progname);
+	} else {
+		log_set_progname("(libsubid)");
 	}
 
 	if (logfd) {
