diff -up shadow-4.1.4.3/lib/commonio.c.selinux shadow-4.1.4.3/lib/commonio.c
--- shadow-4.1.4.3/lib/commonio.c.selinux	2011-02-13 18:58:23.000000000 +0100
+++ shadow-4.1.4.3/lib/commonio.c	2011-11-16 17:09:46.615837659 +0100
@@ -921,10 +921,10 @@ int commonio_close (struct commonio_db *
 
 #ifdef WITH_SELINUX
 	if (db->scontext != NULL) {
-		if (NULL != old_context) {
 		if (setfscreatecon (old_context) < 0) {
 			errors++;
 		}
+		if (NULL != old_context) {
 			freecon (old_context);
 			old_context = NULL;
 		}
