--- shadow-4.0.18.1/src/usermod.c.appendOption	2007-01-14 14:04:29.000000000 +0100
+++ shadow-4.0.18.1/src/usermod.c	2007-01-14 14:05:04.000000000 +0100
@@ -905,7 +905,7 @@
 		 */
 		int c;
 		static struct option long_options[] = {
-			{"append", required_argument, NULL, 'a'},
+			{"append", no_argument, NULL, 'a'},
 			{"comment", required_argument, NULL, 'c'},
 			{"home", required_argument, NULL, 'd'},
 			{"expiredate", required_argument, NULL, 'e'},
